### PR TITLE
Fix email uniqueness validations with oauth

### PR DESCRIPTION
**Pull request by ewjordan**
The User::from_oauth call creates a scope at https://github.com/code-dot-org/code-dot-org/blob/59bcb4528ef6f84f186f03234cd4a04c786e7065/dashboard/app/models/user.rb#L608-L608 which was causing our manual email uniqueness checks to always pass, even when existing users existed in the database. This PR replaces the User table lookups with User.unscoped lookups to clear that scope, and adds a set of tests that catch the problem (they failed without the code change, and now pass).

This only became apparent when looking at Clever account creation with preexisting emails, but it's just a lucky coincidence that things were not broken for Google/FB/MS creations as well - those *always* fail validation because they deliberately do not specify user type (whereas Clever populates it as "student" by default), so email uniqueness is checked outside of the block instead of within it (either in the check for automatic takeover, or after the user selects user type).

### CODE REVIEW

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb [hunk 1]
```diff
@@ -249,4 +249,79 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
...
   end
+
+  # The following tests actually test the user model, but relate specifically to
+  # oauth uniqueness checks so they are included here. These have not been working
+  # in the past for subtle reasons.
+
+  test "omniauth student is checked for email uniqueness against student" do
+    email = 'duplicate@email.com'
+    original_user = create(:user, email: email)
+    assert_empty original_user.email
```
Comment (aoby) [inline dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: Is this assert needed here? This empty student email functionality is already [tested](https://github.com/code-dot-org/code-dot-org/blob/d03519d1a4977b5f7c6abfb5086d25edfef519dd/dashboard/test/models/user_test.rb#L641-L644) in the user model.

Comment (ewjordan) [reply dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: Not particularly, I was mostly just satisfying myself that I was testing what I thought I was testing. Happy to remove.

**Comment by aoby** [comment]
Presumably these users were still failing to create due to the uniqueness check in User itself? If not, we should fix that and add a test.

**Comment by wjordan** [comment]
Is it possible to use `find_or_create_by` instead of `first_or_create` with the `.unscoped` workaround? If so it would make the code a bit  more straightforward.

It seems that `.where.first_or_create` was soft-deprecated ('nodoc'ed - removed from documentation/guides, but without a real deprecation path) in favor of `.find_or_create_by` back in 2012, because scoping has a bunch of open bugs that makes it difficult to use without leading to unexpected behavior like this.

**Comment by wjordan** [comment]
Here's the Rails issue corresponding to this scoping bug in `first_or_[initialize|create]` blocks: rails/rails#17577

**Comment by ewjordan** [comment]
I'm actually very much in favor of that change, since it will minimize the possibility of running into something similar in the future, but not before I have more substantial test coverage in place - I'm really afraid of breaking stuff, our coverage is spotty at best right now, with zero end-to-end automation. So I'll go with this for the moment, but plan to refactor later.

**Comment by ashercodeorg** [comment]
At a higher level, this seems to be an issue with ActiveRecord. Do we know if this is working as intended or an actual bug? It seems worth filing an issue against ActiveRecord.

**Comment by ewjordan** [comment]
@ashercodeorg I agree this is bizarre behavior, and it looks like there is an issue filed at https://github.com/rails/rails/issues/17577, but it doesn't seem like there's much motion on it. https://github.com/rails/rails/pull/24853 seems similar and was closed with "Scopes should leak to callbacks and blocks, it is expected behavior."

So it might just be one of those things :/

**Pull request merged**